File: cypress/support/pages/admin/ordersPage.js
/**
 * Representa a página de gestão pedidos da aplicação.
 * Encapsula seletores e ações relacionadas ao processo de gestão de pedidos.
 */
class OrdersPage {
  elements = {
    ordersMenuLink: () => cy.contains("a", "Manage Orders"),
    /**
     * Input hidden que contém o orderId.
     * @param {string} orderId - O ID do pedido.
     * @returns {Cypress.Chainable<JQuery<HTMLElement>>}
     */
    _orderIdInput: (orderId) =>
      cy.get(`input[type="hidden"][name="orderid"][value="${orderId}"]`),

    /**
     * Container principal de um pedido específico, encontrado via ID.
     * @param {string} orderId - O ID do pedido.
     * @returns {Cypress.Chainable<JQuery<HTMLElement>>}
     */
    orderContainer: (orderId) => {
      const containerSelector = "article.order-item";
      return this.elements._orderIdInput(orderId).closest(containerSelector);
    },

    /**
     * Dropdown de status DENTRO de um container de pedido específico.
     * @param {JQuery<HTMLElement>} orderContainerElement - O elemento container do pedido (obtido via alias).
     * @returns {Cypress.Chainable<JQuery<HTMLSelectElement>>}
     */
    statusSelect: (orderContainerElement) =>
      cy.wrap(orderContainerElement).find('select[name="status"]'),

    /**
     * Botão 'Update' DENTRO de um container de pedido específico.
     * @param {JQuery<HTMLElement>} orderContainerElement - O elemento container do pedido (obtido via alias).
     * @returns {Cypress.Chainable<JQuery<HTMLButtonElement>>}
     */
    updateButton: (orderContainerElement) =>
      cy
        .wrap(orderContainerElement)
        .find('btn btn-alt, button:contains("Update")'),

    /**
     * Badge/Display de status DENTRO de um container de pedido específico.
     * @param {JQuery<HTMLElement>} orderContainerElement - O elemento container do pedido (obtido via alias).
     * @returns {Cypress.Chainable<JQuery<HTMLElement>>}
     */
    statusBadge: (orderContainerElement) =>
      cy.wrap(orderContainerElement).find("span.badge"),
  };

  // --- Ações ---

  /** Navega para a página de gerenciamento de pedidos. */
  navigateToOrdersPage() {
    this.elements.ordersMenuLink().should("be.visible").click();
    cy.url().should("include", "/admin/orders");
    return this;
  }

  /**
   * Localiza o container do pedido pelo ID e o armazena no alias '@currentOrder'.
   * Nota: Este método configura o contexto (alias) para ações subsequentes.
   * @param {string} orderId - O ID do pedido a ser localizado.
   */
  findAndAliasOrderContainer(orderId) {
    this.elements.orderContainer(orderId).as("currentOrder");
    cy.get("@currentOrder").should("exist").and("be.visible");
  }

  /**
   * Seleciona um novo status no dropdown do pedido atualmente em foco (alias @currentOrder).
   * @param {string} status - O texto da opção a ser selecionada (ex: "Fulfilled").
   */
  selectNewStatusForCurrentOrder(status) {
    cy.get("@currentOrder").then((containerElement) => {
      this.elements
        .statusSelect(containerElement)
        .should("be.visible")
        .select(status);
    });
    return this;
  }

  /**
   * Clica no botão "Update" do pedido atualmente em foco (alias @currentOrder).
   * A espera pela requisição (intercept/wait) deve ser feita no step definition.
   */
  clickUpdateForCurrentOrder() {
    cy.get("@currentOrder").then((containerElement) => {
      this.elements.updateButton(containerElement).should("be.visible").click();
    });
    return this;
  }

  // --- Asserções ---

  /**
   * Verifica se o badge de status dentro do pedido em foco (@currentOrder) exibe o texto esperado.
   * @param {string} expectedStatus - O texto do status esperado (ex: "Fulfilled").
   */
  assertOrderStatusForCurrentOrder(expectedStatus) {
    cy.get("@currentOrder").then((containerElement) => {
      this.elements
        .statusBadge(containerElement)
        .should("be.visible")
        .and("contain.text", expectedStatus);
    });
    return this;
  }
}

export default new OrdersPage();

File: pages/orders_page.py
import re

from playwright.sync_api import Locator, Page, expect


class OrdersPage:
    def __init__(self, page: Page):
        self.page = page
        self.header = page.locator("#main-header")
        self.orders_menu_link = self.header.get_by_role("link", name="Manage Orders")
        self.current_order: Locator | None = None

    # --- Elementos (parametrizados) ---

    def _order_container(self, order_id: str):
        hidden_input = self.page.locator(
            f'input[type="hidden"][name="orderid"][value="{order_id}"]'
        )
        return self.page.locator("article.order-item").filter(has=hidden_input)

    @staticmethod
    def _status_select(order_container: Locator):
        return order_container.locator('select[name="status"]')

    @staticmethod
    def _update_button(order_container: Locator):
        return order_container.get_by_role("button", name="Update")

    @staticmethod
    def _status_badge(order_container: Locator):
        return order_container.locator("span.badge")

    # --- Ações ---

    def navigate_to_orders_page(self):
        expect(self.orders_menu_link).to_be_visible()
        self.orders_menu_link.click()
        expect(self.page).to_have_url(re.compile(r".*/admin/orders"))
        return self

    def find_and_focus_order_container(self, order_id: str):
        container = self._order_container(order_id)
        expect(container).to_be_visible()
        self.current_order = container
        return self

    def select_new_status_for_current_order(self, status: str):
        select = self._status_select(self.current_order)
        expect(select).to_be_visible()
        select.select_option(label=status)
        return self

    def click_update_for_current_order(self):
        button = self._update_button(self.current_order)
        expect(button).to_be_visible()
        with self.page.expect_response(re.compile(r"/admin/orders/")):
            button.click()
        return self

    # --- Asserções ---

    def assert_order_status_for_current_order(self, expected_status: str):
        badge = self._status_badge(self.current_order)
        expect(badge).to_be_visible()
        expect(badge).to_contain_text(expected_status)
        return self
